capsules/src/dals/interfaces.rs

```diff
@@ -0,0 +1,107 @@
+    /// `AppLoader` implementation.
+    ///
+    /// - `data_buffer` : Buffer which contains the raw received data for the app
+    /// - `len` : Size of data within the buffer
+    /// - `completed` : Flag which is set when current buffer contains last of the data to be sent to the `AppLoader`
+    fn next_buffer(&self, data_buffer: &'static mut[u8], len: u16, completed: bool);
+
+}
+
+/// Implemented by any module which provides an algorithm to decompress the app data buffer-by-buffer
+pub trait Decompressor<'a>{
+    fn set_client(&self,client: &'a dyn DecompressorClient);
+    /// Called by `DecompressorClient` implementation
+    ///
+    /// - `buffer` : Buffer of data awaiting to be decompressed.
+    /// - `len` : Size of data within buffer.
+    fn decompress_buffer(&self, buffer: &'static mut [u8], len: u16) -> Result<(),DecompressionError>;
+}
+
+/// Implemented by main `AppLoader` module
+pub trait DecompressorClient<'a>{
+    /// Called by `Decompressor` implementation
+    ///
+    /// - `decompressed_data_buffer` : Buffer which contains the next sqeuence of decompressed app data.
+    /// - `original_data_buffer` : Original buffer passed via `decompress_buffer().  This original buffer is
+    /// passed back to `DecompressorClient` only when all of the data in it has been decompressed.
+    /// In other words, this buffer being passed back indicates to the `DecompressorClient` that all of the data in it
+    /// has been processed by the `Decompressor`
+    fn decompressed_buffer(&self, decompressed_data_buffer: &'static [u8], len: u16, original_data_buffer: Option<&'static [u8]>);
+}
+
+
+/// Implemented by any module which provides an algorithm to verify the loaded app binary for security
+/// purposes (ex: SHA, Checksum, MD5 implementations)
+pub trait Verifier<'a>{
+    fn set_client(&self, client: &'a dyn VerifierClient);
+    /// Called by `VerifierClient` implementation
+    ///
+    /// - `app_flash` : Immutable reference to app data stored in flash
+    /// - `len` : Size of app in flash
+    fn verify_data(&self, app_flash: &'static [u8], len: u32); 
+}
+
+/// Implemented by main `AppLoader` module
+pub trait VerifierClient<'a>{
+    /// Called by `Verifier` implementation
+    ///
+    /// - `Option<VerificationError>` : Optional error passed back to indicate why verification failed.  No error returned if verification is successful.
+    fn verification_complete(&self, error: Option<VerificationError>); 
+}
+
+/// Implemented by any module which aims to validate the loaded app according to user specifications 
+pub trait Validator<'a>{
+    fn set_client(&self, client: &'a dyn ValidatorClient);
+    /// Called by `ValidatorClient`
+    ///
+    /// - `app_flash` : Immutable reference to app data stored in flash
+    /// - `len` : Size of app in flash 
+    fn validate_data(&self, app_flash: &'static [u8], len: u32);
```

Comment:
By handing the Validator a `&'static [u8]`, the AppLoader loses the ability to modify that data in flash until the system is restarted. Is that intended?

Reply:
Implicitly, yes, as in the current design AppLoader would have no need to write the flash at this point. The app has already been entirely loaded.

Reply:
This suggests that once an app has been loaded into a location it cannot be removed and replaced by another app without a system restart?

Reply:
Yes.

Reply:
hmm, okay. I imagined that DALS would also enable app updates by making it possible to replace an existing app with a new version, but I suppose that will require some other mechanism.

Reply:
I suppose a slight change to the interface could be made where Verifier/Validator was pass a reference to the flash region back and forth.  This would enable the AppLoader to receive this pointer again after the Verification and Validation stages are over and make any necessary changes to the app.

Reply:
The core kernel can create any slices it wants, so perhaps we want the core kernel to provide the slice when it allocates space for DALS to store the app.

In general, the interface to flash is not specified in this PR, but I'm imagining that AppLoader will have to ask the core kernel where it can store an incoming app of a certain length, and the core kernel will provide some mechanism for DALS to write the app in that location.